File: #1642.furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& heights, int bricks, int ladders) {

        int size = heights.size();
        priority_queue<int> pq;
        int i = 1;
        for (; i < size; i++) {
            int d = heights[i] - heights[i - 1];
            if (d <= 0) {
                continue;
            } else if (d <= bricks) {
                bricks -= d;
                pq.push(d);
            } else if (!pq.empty() && ladders && pq.top() > d) {
                bricks += pq.top() - d;
                pq.pop();
                --ladders;
                pq.push(d);
            } else if (ladders) {
                --ladders;
            } else {
                break;
            }
        }
        return i - 1;
    }
};

File: #1642.furthest-building-you-can-reach.js
/**
 * @param {number[]} heights
 * @param {number} bricks
 * @param {number} ladders
 * @return {number}
 */
var furthestBuilding = function(heights, bricks, ladders) {
    const pq = new MinPriorityQueue();
    for (let i = 0; i < heights.length - 1; i++) {
        let diff = heights[i+1] - heights[i];
        if (diff < 0) {
            continue;
        }
        
        pq.enqueue(diff);
        if (pq.size() > ladders) {
            bricks -= pq.dequeue().element;
            if (bricks < 0) {
                return i;
            }
        }
    }
    return heights.length - 1;
};
